### PR TITLE
[ucd/age] Change Age type to struct

### PR DESCRIPTION
**Pull request by behnam**
The *Age* property was implemented as an `enum` type which was closely
mirroring the core `Option` type. Now that we have a clear definition of
`Partial` properties, it makes sense to update `Age` to be a Custom
property type (`struct`) with *actual* `UnicodeVersion` value.

But, there was a need to update one thing about the Age, it's ordering!
Previously, Age was ordered in the same direction as `UnicodeVersion`,
and `Unassigned` value was larger than any `Assigned(UnicodeVersion)`.
That didn't make much sense in common understanding of *age*, but was
consistent with `UnicodeVersion`.

Now, because of the fact that we cannot change the ordering of `Option`
variants (`None` and `Some(UnicodeVersion)`, in this case), and also to
put the meaning of comparison closer to the common understanding of the
word, we reverse the direciton of ordering for `Age`. Therefore, `Age`
value for character `'A'` is *greater than or equal to* any other
character in Unicode, assigned or unassigned.

Closes <https://github.com/behnam/rust-unic/issues/120>

### CODE REVIEW

**Comment by CAD97** [comment]
I would like to say that having `Unassigned > Assigned` is ideal.

<http://unicode.org/reports/tr44/#Character_Age>

> The typical use case for the Age property in regular expressions is to search for all characters that were present in a given version. For this reason, an expression such as "\p{age=V3_0}" is exceptionally defined to match all of the code points assigned in Version 3.0—that is, all the code points with a value less than or equal to the value 3.0 for the Age property, rather than just the subset of those code points with the value 3.0. This interprets "\p{age=V3_0}" as the set of all characters assigned as of Unicode 3.0, rather than as just the set of characters added to Unicode 3.0 subsequent to the prior version. For more information, see Unicode Technical Standard #18, "Unicode Regular Expressions" [[UTS18]].

So the filter for this behavior is `Age::of(char) <= Age(UnicodeVersion::parse("3.0.0")` if Unassigned > Assigned.

The key point here is to realize that the `Age` property is not some how long the character has been around, but the version in which the codepoint was assigned.

[UTS18]: http://unicode.org/reports/tr41/tr41-21.html#UTS18

**Comment by CAD97** [comment]
That said, (and see #120 for much more said), I feel using `Option` rather than re-implementing the concept is best. Anybody programming in Rust already should understand the ordering of `Option::None` versus `Option::Some` as well, so this avoids surprises there by using the stdlib type.

File: unic/ucd/normal/src/canonical_combining_class.rs [hunk 1]
```diff
@@ -53,7 +53,7 @@ impl TotalCharProperty for CanonicalCombiningClass {
 impl NumericCharProperty<u8> for CanonicalCombiningClass {
     /// Get numeric value for character property value
     fn number(&self) -> u8 {
-        self.number()
+        Self::number(self)
```
Comment (CAD97) [inline unic/ucd/normal/src/canonical_combining_class.rs]: Wait. How does this work? Is this not self-recursive and should it not just be `self.0`?

File: gen/src/generate/ucd/age.rs [hunk 2]
```diff
@@ -74,7 +74,7 @@ impl FromStr for AgeData {
                     age_data.insert(
                         char,
                         format!(
-                            "Assigned(UnicodeVersion {{ major: {}, minor: {}, micro: {} }})",
+                            "UnicodeVersion {{ major: {}, minor: {}, micro: {} }}",
```
Comment (CAD97) [inline gen/src/generate/ucd/age.rs]: This feels so much nicer. 🎉

Comment (behnam) [inline gen/src/generate/ucd/age.rs]: Yeah! Actually, I think we should do the same for all non-named properties. Doing so, we don't have to make the struct fields public at all!

File: unic/ucd/age/src/age.rs [hunk 3]
```diff
@@ -10,36 +10,37 @@
...
-/// The *earliest* value for this property is `UnicodeVersion { major: 1, minor: 1, micro: 0 }`,
-/// because of the massive changes for the merger of the Unicode Stanrda with ISO 10646.
+/// Unicode versions with character *assignement* always have the Micro (Update) version value
+/// of zero (`0`). Therefore, all `UnicodeVersion` values return from `Age` will have their `macro`
+/// field as `0`.
 ///
-/// The *latest* value for this property is always equal to or less than `UNICODE_VERSION`. (Only
-/// not equal when `UNICODE_VERSION` has non-zero *micro* value.)
+/// The *earliest* (largest) value for this property is `UnicodeVersion { major: 1, minor: 1, micro:
+/// 0 }`, because of the massive changes for the merger of the Unicode Stanrda with ISO 10646.
```
Comment (behnam) [inline unic/ucd/age/src/age.rs]: Thanks! Fixed here and in other copy.

File: unic/ucd/normal/src/canonical_combining_class.rs (hunk 1 above)
Comment (behnam) [inline unic/ucd/normal/src/canonical_combining_class.rs]: If a type has a method, that always takes precedence. So, calling into such a method from a trait is safe, as long as you remember to implement the function. With the right lint warn enabled, we can easily catch the problem in compile time, if happens. (Which actually did while I was expanding the contracts! :D)

**Comment by behnam** [comment]
bors: r+

**Pull request merged**